pkg/windows/service/service.go

```diff
@@ -80,9 +79,19 @@ Loop:
 			case svc.Interrogate:
 				s <- c.CurrentStatus
 			case svc.Stop, svc.Shutdown:
-				s <- svc.Status{State: svc.Stopped}
-				// TODO: Stop the kubelet gracefully instead of killing the process
-				os.Exit(0)
+				klog.Infof("Service stopping")
+				// We need to translate this request into a signal that can be handled by the the signal handler
+				// handling shutdowns normally (currently apiserver/pkg/server/signal.go).
+				// If we do not do this, our main threads won't be notified of the upcoming shutdown.
+				// Since Windows services do not use any console, we cannot simply generate a CTRL_BREAK_EVENT
+				// but need a dedicated notification mechanism.
+				server.RequestShutdown()
```

Comment:
i am not a golang expert, but why is this needed? i would think that `break loop` will allow the process to gracefully exit after you set the status to `stopPending`

Reply:
by this, i mean the extra complexity around `server.RequestShutdown`

Reply:
@michmike 
As the comments describe:
- The thread/goroutine running this communicates with the windows service controller.
- `break loop` will only exit that one thread, while the main thread (e.g. kubelet) will continue to run and
eventually be (forcefully) timeout-killed by the service controller.
By doing that we'd also essentially communicate to the service controller: "We're currently shutting down" but we'll never actually shut down, which it doesn't like.

- RequestShutdown() basically executes the same code that would've been executed if a SIGINT signal was received (which you cannot send/receive in "windows service mode". In windows that doesn't actually exist but there's CTRL_BREAK/CTRL_C event which are mapped to signals like SIGINT by golang. Those cannot be sent to processes without a console (User I/O) - which a service by definition does not have. So this is the least-intrusive approach to add a alternative shutdown-signaling mechanism.
- I also wouldn't call it complexity - the alternative (assuming a parallel universe where this would work) is sending the signal manually, which is much more complex than adding a single function, that might be of use elsewhere.

Let me know if that clarifies things and what you think